### PR TITLE
Fix #4440: Do not serialize the content of static objects

### LINKED ISSUE #4440: Serializable objects should have a "readResolve" method synthesized
See explanation in https://github.com/scala/scala/blob/a9275b50e425d29c0b7fd45c10317953d5bf7804/src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala#L319-L324
In Dotty this should be done in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/SyntheticMethods.scala

### PR DESCRIPTION
**Pull request by smarter**
In #4450 and Scala 2.12, readResolve is used to make sure deserializing
an object returns the singleton instance of the object, but this doesn't
prevent the fields of the objects from being serialized in the first
place even though they're not used. Scala 2.13 switched to using
writeReplace to completely bypass serialization of the object in
https://github.com/scala/scala/pull/7297. This commit adapts this to
Dotty.

Co-Authored-By: Ingar Abrahamsen <ingar@abrahams1.com>
Co-Authored-By: Jason Zaugg <jzaugg@gmail.com>

### CODE REVIEW

**Comment by smarter** [comment]
@retronym feel free to review and let us if there are relevant test cases from scalac we should add.

**Comment by smarter** [comment]
@nicolasstucki I just added support for `classOf[X.type]` where `X` is an object but this breaks the decompilation tests because it gets printed as `classOf[X]` and I can't figure out how to fix that, can you have a look ?

**Comment by nicolasstucki** [comment]
I will have a look

File: library/src/scala/tasty/reflect/Printers.scala [hunk 1]
```diff
@@ -646,6 +646,7 @@ trait Printers
                   n == "copy" ||
                   n.matches("copy\\$default\\$[1-9][0-9]*") || // default parameters for the copy method
                   n.matches("_[1-9][0-9]*") || // Getters from Product
+                  n == "writeReplace" || // Case class serialization
```
Comment (smarter) [inline library/src/scala/tasty/reflect/Printers.scala]: That's not correct: the user is free to write their own writeReplace method (e.g. `Baz` in https://github.com/lampepfl/dotty/blob/bb3673cd3aec19cfe01494698d62c70aed496e20/tests/run/serialize.scala)

Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/Printers.scala]: It does not print the code only if the method is synthetic. The comment is misleading. I will update it.

Comment (smarter) [reply library/src/scala/tasty/reflect/Printers.scala]: OK, we still need to properly print `classOf[X.type]` to decompile https://github.com/lampepfl/dotty/blob/91574bd6689518282437eee6f6a31bb287cf377d/tests/run/classof-object.scala :).

Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/Printers.scala]: Yes, we do but I want to merge the last PR from the decompiler project before doing any changes.

**Comment by nicolasstucki** [comment]
Now the tests pass

File: library/src/scala/tasty/reflect/Printers.scala [hunk 2]
```diff
@@ -637,21 +637,22 @@ trait Printers
...
               // Currently the compiler does not allow overriding some of the methods generated for case classes
               d.symbol.flags.is(Flags.Synthetic) &&
               (d match {
-                case DefDef("apply" | "unapply", _, _, _, _) if d.symbol.owner.flags.is(Flags.Object) => true
+                case DefDef("apply" | "unapply" | "writeReplace", _, _, _, _) if d.symbol.owner.flags.is(Flags.Object) => true
                 case DefDef(n, _, _, _, _) if d.symbol.owner.flags.is(Flags.Case) =>
                   n == "copy" ||
                   n.matches("copy\\$default\\$[1-9][0-9]*") || // default parameters for the copy method
                   n.matches("_[1-9][0-9]*") || // Getters from Product
+                  n == "writeReplace" || // Case class serialization
```
Comment (OlivierBlanvillain) [inline library/src/scala/tasty/reflect/Printers.scala]: I think the comment (and method name) is miss leading given that `writeReplace` is added on all objects, not just case objects.

File: compiler/src/dotty/tools/dotc/transform/SyntheticMethods.scala [hunk 3]
```diff
@@ -23,8 +23,9 @@ import ValueClasses.isDerivedValueClass
  *    def productArity: Int
  *    def productPrefix: String
  *
- *  Special handling:
- *    protected def readResolve(): AnyRef
```
[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/SyntheticMethods.scala]: This was the previous mechanism used in scalac and never implemented in dotty to do object serialization, I'll remove it from StdNames.

File: library/src/scala/tasty/reflect/Printers.scala (hunk 2 above)
Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/Printers.scala]: Actually, the previous case should be the one that handles the `writeReplace`. I will remove this line.

**Pull request merged**